internal/jwtparser/jwt_test.go

```diff
@@ -0,0 +1,203 @@
+			return
+		}
+		hdrs := w.Header()
+		hdrs.Set(`Content-Type`, `application/json`)
+		pub, _ := s.keySet.Key(0)
+		_ = json.NewEncoder(w).Encode(pub)
+	}))
+
+	s.url = s.srv.URL + "/.well-known/jwks.json"
+}
+
+func (s *jwtParserSuite) TearDownTest(_ *gc.C) {
+	s.srv.Close()
+}
+
+func HTTPClient() *http.Client {
+	return &http.Client{
+		Timeout: 5 * time.Second,
+	}
+}
+
+func (s *jwtParserSuite) TestCacheRegistration(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), s.url)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *jwtParserSuite) TestCacheRegistrationFailureWithBadURL(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), "noexisturl")
+	// We want to make sure that we get an error for a bad url.
+	c.Assert(err, gc.NotNil)
+}
+
+func (s *jwtParserSuite) TestParseJWT(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), s.url)
+	c.Assert(err, jc.ErrorIsNil)
+
+	params := JWTParams{
+		audience: "controller-1",
+		subject:  "alice",
+		claims:   map[string]string{"model-1": "read"},
+	}
+	jwt, err := EncodedJWT(params, s.keySet, s.signingKey)
+	c.Assert(err, jc.ErrorIsNil)
+	base64jwt := base64.StdEncoding.EncodeToString(jwt)
+
+	token, err := authenticator.Parse(context.Background(), base64jwt)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(token, gc.NotNil)
+
+	claims := token.PrivateClaims()
+	c.Assert(token.Subject(), gc.Equals, "alice")
+	c.Assert(token.Issuer(), gc.Equals, "test")
+	c.Assert(token.Audience(), jc.DeepEquals, []string{"controller-1"})
+	c.Assert(token.Expiration().After(token.IssuedAt()), gc.Equals, true)
+	c.Assert(claims["access"], jc.DeepEquals, map[string]interface{}{"model-1": "read"})
+}
+
+func (s *jwtParserSuite) TestSimultaneousParse(c *gc.C) {
+	ctx, done := context.WithCancel(context.Background())
+	defer done()
+	authenticator := jwtparser.NewParserWithHTTPClient(ctx, HTTPClient())
+	err := authenticator.SetJWKSCache(context.Background(), s.url)
+	c.Assert(err, jc.ErrorIsNil)
+
+	params := JWTParams{
+		audience: "controller-1",
+		subject:  "alice",
+		claims:   map[string]string{"model-1": "read"},
+	}
+	jwt, err := EncodedJWT(params, s.keySet, s.signingKey)
+	c.Assert(err, jc.ErrorIsNil)
+	base64jwt := base64.StdEncoding.EncodeToString(jwt)
+
+	n := 10
+	for range n {
+		go func() {
+			token, err := authenticator.Parse(context.Background(), base64jwt)
+			c.Check(err, jc.ErrorIsNil)
+			c.Check(token, gc.NotNil)
+		}()
+	}
+}
+
+// NewJWKSet returns a new key set and signing key.
+func NewJWKSet() (jwk.Set, jwk.Key, error) {
```

Comment:
nit: a pattern we tend to do that you may find useful in tests. When we have a func that does some utility work and has the potential we pass the c of the test in so the func just do the error checks. Makes the tests a bit easier to read and you don't have to check the error then.

Up to you.